File: limitlion/throttle.lua
-- This script implements a per second token bucket rate limiting algorithm.  It is
-- based on Stripe's published script.
-- KEYS = {}  ARGV = { throttle knob name, default rate (rps), default burst multiplier,
--                     default rate limit window in seconds, requested tokens, knobs_ttl }
-- Returns: allowed (1=allowed, 0=not allowed),
--          tokens left,
--          decimal seconds left in this window

-- Using the time command requires this script to be replicated via commands
redis.replicate_commands()

local function check_bucket(bucket_key, rate, burst, window,
                            now, requested_tokens)
  -- Checks bucket to see if a request would be allowed.
  --
  -- Args:
  --   bucket_key: Redis key name of bucket
  --   rate: Request rate per second
  --   burst: Burst multiplier
  --   window: Number of seconds in window
  --   now: Current second since epoch
  --   requested_tokens: Number of tokens requested
  --
  -- Returns:
  --   allowed: 1 if this request should allowed, otherwise 0
  --   refreshed: Window start time, whole seconds since epoch
  --   filled_tokens: How many tokens in bucket

  -- Maximum size of bucket
  local capacity = math.ceil(rate * burst * window)

  local last_tokens = tonumber(redis.call("hget", bucket_key, "tokens"))
  if last_tokens == nil then
    last_tokens = capacity
  end

  local last_refreshed = tonumber(redis.call("hget", bucket_key, "refreshed"))
  if last_refreshed == nil then
    last_refreshed = 0
  end

  -- Calculate how many new tokens should be added, can be zero
  local age = math.max(0, now-last_refreshed)
  -- Whole windows that have elapsed
  local elapsed_windows = math.floor(age / window)
  local add_tokens = math.ceil(elapsed_windows * rate * window)

  -- Fill bucket with new tokens
  local filled_tokens = math.min(capacity, last_tokens + add_tokens)

  -- Determine if this request is going to be allowed
  local allowed
  if filled_tokens >= requested_tokens then
    allowed = 1
  else
    allowed = 0
  end

  local refreshed
  if add_tokens > 0 and last_refreshed == 0 then
    -- Adding tokens to a new bucket.
    refreshed = now
  elseif add_tokens > 0 and last_refreshed ~= 0 then
    -- Add tokens to an existing bucket.
    refreshed = last_refreshed + elapsed_windows * window
  else
    -- Don't change refreshed time if we haven't added new tokens
    refreshed = last_refreshed
  end

  return {allowed, refreshed, filled_tokens}
end

local function update_bucket(bucket_key, allowed, refreshed,
                             filled_tokens, ttl, requested_tokens)
  -- Updates bucket token count, last refreshed time, and TTL
  --
  -- Args:
  --   bucket_key: Redis key name of bucket
  --   allowed: 1 if this request will be allowed, otherwise 0
  --   refreshed: Window start time, whole seconds since epoch
  --   filled_tokens: How many tokens in bucket
  --   ttl: Redis key expiration
  --
  -- Returns:
  --   new_tokens: Current number of tokens in bucket

  local new_tokens = filled_tokens
  if allowed == 1 then
    new_tokens = math.max(0, filled_tokens - requested_tokens)
  end

  redis.call("hmset", bucket_key, "tokens", new_tokens, "refreshed", refreshed)
  redis.call("expire", bucket_key, ttl)
  return new_tokens
end

local name = ARGV[1]
local default_rps = ARGV[2]
local default_burst = ARGV[3]
local default_window = ARGV[4]
local requested_tokens = tonumber(ARGV[5])
local knobs_ttl = tonumber(ARGV[6])
local rps
local burst
local window

-- Lookup throttle knob settings
local knobs_key = name .. ":knobs"
-- Use
--   HMSET <knobs_key> rps <rps> burst <burst> window <window>
-- to manually override the setting for any throttle.
local knobs = redis.call("HMGET", knobs_key, "rps", "burst", "window")
if knobs[1] == false then
  -- Use defaults if knobs hash is not found
  rps = tonumber(default_rps)
  burst = tonumber(default_burst)
  window = tonumber(default_window)
else
  rps = tonumber(knobs[1])
  burst = tonumber(knobs[2])
  window = tonumber(knobs[3])
  -- Set knobs hash expiration if knobs_ttl is specified
  if knobs_ttl > 0 then
    redis.call("EXPIRE", knobs_key, knobs_ttl)
  end
end

-- Use redis server time so it is consistent across callers
-- The following line gets replaced before loading this script during tests
-- so that we can freeze the time values.  See throttle.py.
local time = redis.call("time")
local now = tonumber(time[1])

-- Keep the hash around for twice the useful burst time to reduce unnecessary expires
local ttl = math.floor(burst * window * 2)

local tokens
local seconds_left
local allowed
if rps == 0 then
  -- rps = 0 always results in a denied request with a full window sleep
  seconds_left = window
  tokens = 0
  allowed = 0
elseif rps == -1 then
  -- rps = -1 always results in an allowed
  seconds_left = 0
  tokens = 1
  allowed = 1
else
  -- Check bucket to determine if work is allowed
  local rate = check_bucket(name, rps, burst, window, now, requested_tokens)
  tokens = update_bucket(name, rate[1], rate[2], rate[3], ttl, requested_tokens)
  allowed = rate[1]
  -- Calculate decimal seconds left in the window
  local diff = math.max(0, now - tonumber(rate[2]))
  seconds_left = (window - diff - 1) + (1000000 - tonumber(time[2])) / 1000000
end

-- string.format is necessary for seconds_left because Lua to Redis number
-- conversion automatically casts numbers to integers which would drop the microseconds
return {allowed, tokens, string.format("%.6f", seconds_left)}

File: limitlion/throttle.py
"""Token bucket throttle backed by Redis."""

import time

import pkg_resources

KEY_FORMAT = 'throttle:{}'

# throttle knob defaults
THROTTLE_BURST_DEFAULT = 1
THROTTLE_WINDOW_DEFAULT = 5
THROTTLE_REQUESTED_TOKENS_DEFAULT = 1

# The default is to extend a throttle's knob settings TTL out
# 7 days each time the throttle is used.
DEFAULT_KNOBS_TTL = 60 * 60 * 24 * 7

throttle_script = None
redis = None


def _validate_throttle(key, params):
    check_values_pipe = redis.pipeline()
    for param, param_name in params:
        if param is not None:
            # Throttle values can only be positive floats
            try:
                assert float(param) >= 0
            except (ValueError, AssertionError):
                raise ValueError(
                    '"{}" is not a valid throttle value. Throttle values must '
                    'be positive floats.'.format(param)
                )
        else:
            check_values_pipe.hexists(key, param_name)
    if not all(check_values_pipe.execute()):
        raise IndexError(
            "Throttle knob {} doesn't exist or is invalid".format(key)
        )


def _verify_configured():
    if not redis or not throttle_script:
        raise RuntimeError('Throttle is not configured')


def throttle(
    name,
    rps,
    burst=THROTTLE_BURST_DEFAULT,
    window=THROTTLE_WINDOW_DEFAULT,
    requested_tokens=THROTTLE_REQUESTED_TOKENS_DEFAULT,
    knobs_ttl=DEFAULT_KNOBS_TTL,
):
    """
    Throttle that allows orchestration of distributed workers.

    Args:
        name: Name of throttle.  Used as part of the Redis key.
        rps: Default requests per second allowed by this throttle
        burst: Default burst multiplier
        window: Default limit window in seconds
        requested_tokens: Number of tokens required for this work request
        knobs_ttl: Throttle's knob TTL value (0 disables setting TTL)

    Returns:
        allowed: True if work is allowed
        tokens: Number of tokens left in throttle bucket
        sleep: Seconds before next limit window starts.  If work is
               not allowed you should sleep this many seconds. (float)

    The first use of a throttle will set the default values in redis for
    rps, burst, and window. Subsequent calls will use the values stored in
    Redis. This allows changes to the throttle knobs to be made on the fly by
    simply changing the values stored in redis.

    See throttle_set function to set the throttle.

    Setting RPS to 0 causes all work requests to be denied and a full sleep.
    Setting RPS to -1 causes all work requests to be allowed.

    """

    _verify_configured()
    allowed, tokens, sleep = throttle_script(
        keys=[],
        args=[
            KEY_FORMAT.format(name),
            rps,
            burst,
            window,
            requested_tokens,
            knobs_ttl,
        ],
    )
    # Converting the string sleep to a float causes floating point rounding
    # issues that limits having true microsecond resolution for the sleep
    # value.
    return allowed == 1, int(tokens), float(sleep)


def throttle_configure(redis_instance, testing=False):
    """Register Lua throttle script in Redis."""

    global redis, throttle_script
    redis = redis_instance

    lua_script = pkg_resources.resource_string(
        __name__, 'throttle.lua'
    ).decode()

    # Modify scripts when testing so time can be frozen
    if testing:
        lua_script = lua_script.replace(
            'local time = redis.call("time")',
            'local time\n'
            'if redis.call("exists", "frozen_second") == 1 then\n'
            '  time = redis.call("mget", "frozen_second", "frozen_microsecond")\n'  # noqa: E501
            'else\n'
            '  time = redis.call("time")\n'
            'end',
        )
    throttle_script = redis.register_script(lua_script)


def throttle_delete(name):
    """Delete Redis throttle data."""

    _verify_configured()
    key = KEY_FORMAT.format(name)
    pipeline = redis.pipeline()
    pipeline.delete(key)
    pipeline.delete(key + ':knobs')
    pipeline.execute()


def throttle_get(name):
    """
    Get throttle values from redis.

    Returns: (tokens, refreshed, rps, burst, window)

    """

    key = KEY_FORMAT.format(name) + ':knobs'

    # Get each value in hashes individually in case they don't exist
    get_values_pipe = redis.pipeline()
    key = KEY_FORMAT.format(name)
    get_values_pipe.hget(key, 'tokens')
    get_values_pipe.hget(key, 'refreshed')

    key = KEY_FORMAT.format(name) + ':knobs'
    get_values_pipe.hget(key, 'rps')
    get_values_pipe.hget(key, 'burst')
    get_values_pipe.hget(key, 'window')

    values = get_values_pipe.execute()
    return values


def throttle_reset(name):
    """Reset throttle settings."""

    _verify_configured()
    key = KEY_FORMAT.format(name) + ':knobs'
    redis.delete(key)


def throttle_set(name, rps=None, burst=None, window=None, knobs_ttl=None):
    """
    Adjust throttle values in redis.

    If knobs_ttl is used here the throttle() call needs to be called
    with knobs_ttl=0 so the ttl isn't also set in the Lua script
    """

    _verify_configured()
    key = KEY_FORMAT.format(name) + ':knobs'

    params = [(rps, 'rps'), (burst, 'burst'), (window, 'window')]
    _validate_throttle(key, params)

    set_values_pipe = redis.pipeline()
    for param, param_name in params:
        if param is not None:
            set_values_pipe.hset(key, param_name, param)

    if knobs_ttl:
        set_values_pipe.expire(key, knobs_ttl)

    set_values_pipe.execute()


def throttle_wait(name, *args, **kwargs):
    """Sleeps time specified by throttle if needed.

    This will wait potentially forever to get permission to do work

    Usage:
    throttle = throttle_wait('name', rps=123)
    for ...:
        throttle()
        do_work()
    """

    max_wait = kwargs.pop('max_wait', None)

    def throttle_func(requested_tokens=1):
        start_time = time.time()
        allowed, tokens, sleep = throttle(
            name, *args, requested_tokens=requested_tokens, **kwargs
        )
        while not allowed:
            if max_wait is not None and time.time() - start_time > max_wait:
                break
            time.sleep(sleep)
            allowed, tokens, sleep = throttle(
                name, *args, requested_tokens=requested_tokens, **kwargs
            )
        return allowed, tokens, sleep

    return throttle_func
